### PR TITLE
Proposal: traited fact serialization

### PR DESCRIPTION
**Pull request by Northburns**
See issue #114 I rambled about this there 😅

### CODE REVIEW

File: src/main/kotlin/io/github/kryszak/gwatlin/http/serializers/JsonUtil.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package io.github.kryszak.gwatlin.http.serializers
+
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+/**
+ * A convenience object for [Json], and to help the
+ * library to use the same instance of it throughout.
+ */
+object JsonUtil {
```
Comment (Kryszak) [inline src/main/kotlin/io/github/kryszak/gwatlin/http/serializers/JsonUtil.kt:10]: I'm not a big fan of `*Util` class naming. In this case the object seems like `JsonConfigurer` maybe? I'm not forcing rename, just thinking out loud :)

File: src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    class TraitedFactUnwrapSerializer : JsonTransformingSerializer<TraitedFact>(serializer()) {
+        private fun split(element: JsonElement) = element.jsonObject.entries
+            .groupBy { (k, _) -> k == "requires_trait" || k == "overrides" }
+            .mapValues { (_, v) -> v.associate { it.toPair() } }
+            .let { it[true] to it[false] }
+
+        override fun transformSerialize(element: JsonElement): JsonElement {
+            val (traitedFact, fact) = split(element)
+            val entries = requireNotNull(traitedFact).toMutableMap()
+            requireNotNull(fact).entries.single().value.jsonObject.entries
```
Comment (Kryszak) [inline src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt:36]: This line would be more readable if split into multiple ones IMO
```suggestion
            requireNotNull(fact)
              .entries
              .single()
              .value
              .jsonObject
              .entries
              .forEach { (k, v) -> entries[k] = v }
```

File: src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    @SerialName("requires_trait")
+    val requiresTrait: Int,
+    val overrides: Int? = null,
+    val fact: Fact,
+) {
+    /**
+     * Use this to serialize [TraitedFact] to comply with GW2 API.
+     */
+    class TraitedFactUnwrapSerializer : JsonTransformingSerializer<TraitedFact>(serializer()) {
+        private fun split(element: JsonElement) = element.jsonObject.entries
```
Comment (Kryszak) [inline src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt:28]: Let's put it on bottom of class, overridden function are more important and should be first thing to see here

**Review by Kryszak — CHANGES_REQUESTED** [review]
Just couple of stylistic comments, main logic looks great!

File: src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt (hunk 2 above)
Comment (Northburns) [reply src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt]: I can agree with this, fixed 😊

File: src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt (hunk 3 above)
Comment (Northburns) [reply src/main/kotlin/io/github/kryszak/gwatlin/api/gamemechanics/model/facts/TraitedFact.kt]: Good point, fixed.

**Review by Kryszak — APPROVED** [review]

**Pull request merged**